### PR TITLE
Add support for Check-Prompt-Resolve (CPR) scripts in arc diff.

### PR DESCRIPTION
**Pull request by lazaroclapp**
Check-Prompt-Resolve (CPR) scripts are custom checks to be added
before unit and lint tests, which allow us to prompt the user to
run a (potentially heavyweight) fix script if a provided
lightweight check fails.

It adds the following optional entry type to `.arcconfig`:

```php
"uber.differential.check_prompt_resolve": [
    {
      "check-script" : "./path/to/quick_check.sh",
      "prompt" : "[Explanation]. Fix it? (This can take a few minutes)",
      "default-to" : true,
      "resolve-script" : "./path/to/fix_script.sh"
    }
  ]
```

For each object in the list, `arc diff` will run the `check-script`
unconditionally. If it succeeds, it will move on to the next CPR or
to lint/unit tests. If the check fails, it will prompt the user with
`prompt` and either [Y/n] (if `default-to` is `true`) or [y/N] (if
`default-to` is `false`). If the user answers yes, it will run
the `resolve-script`.

The reason for this workflow vs `arc lint` autofix, is that `arc lint`
requires the patch to be provided at the same time as the fix. The
use case for CPRs is that of an expensive script needed to fix the
diff, which is rarely run, and only needed if some easy to check
condition is true.

Note that, for repositories not defining 
`uber.differential.check_prompt_resolve`, this PR should make
no changes to arc behavior.

### CODE REVIEW

**Comment by lazaroclapp** [comment]
cc: @msridhar , @kageiit , @raviagarwal7

File: src/workflow/ArcanistDiffWorkflow.php [hunk 1]
```diff
@@ -1187,6 +1189,61 @@ private function shouldAmend() {
...
+      $prompt = $cprDef['prompt'];
+      $defaultsToYes = $cprDef['default-to'];
+      $resolveScript = $cprDef['resolve-script'];
+      list ($err, $stdout, $stderr) = exec_manual('%s', $checkScript);
+      if ($err == 0) {
+        continue; // Nothing to fix
+      }
+      $doResolve = phutil_console_confirm(
+                            $prompt, $default_no = !$defaultsToYes);
+      list ($err, $stdout, $stderr) = exec_manual('%s', $resolveScript);
```
Comment (raviagarwal7) [inline src/workflow/ArcanistDiffWorkflow.php]: probably pipe `$stdout` and `$stderr` to console since these would could include full build logs, etc.

Comment (lazaroclapp) [reply src/workflow/ArcanistDiffWorkflow.php]: 👍

**Comment by lazaroclapp** [comment]
> Does the fix-ups automatically get's added to the commit?

Currently, that's the responsibility of the resolve script. We currently are running `git amend --no-edit && git add` there. I agree that moving that logic inside `arc diff` itself might be better.

> We should prompt users with the fixup being done before applying. similar to how `arc lint` does, to prevent any unwanted changes to user's state.

Lint fixes are usually very small, and thus the developer can easily see the change . This won't be true for many CPR patches. If we really want to prompt the user with something here, a list of affected files might be better than a diff. They can then decide whether or not to amend the commit. Note that, after they elected to run the resolve script, this might leave changes in their local git state, unless we manually run git stash or `git checkout HEAD $f` for each file $f changed by the resolve script.

> Does it run as part of arc land?

Do we care? `arc diff` should always happen before `arc land`, right? Which means the check script will always succeed for `arc land` and there will be nothing to do. Am I misunderstanding the workflow?

> We would need verify that the automatic diff creation works where all the arguments can be passed on the command line

Will add `--nocpr` and make it be implied by `--only` as well.

**Comment by raviagarwal7** [comment]
don't need to run as part of `arc land`. probably just check that the `arc land` and `arc submit` workflows works as expected.

**Comment by lazaroclapp** [comment]
> > > We should prompt users with the fixup being done before applying. similar to how `arc lint` does, to prevent any unwanted changes to user's state.
> > 
> > 
> > Lint fixes are usually very small, and thus the developer can easily see the change . This won't be true for many CPR patches. If we really want to prompt the user with something here, a list of affected files might be better than a diff. They can then decide whether or not to amend the commit. Note that, after they elected to run the resolve script, this might leave changes in their local git state, unless we manually run git stash or `git checkout HEAD $f` for each file $f changed by the resolve script.
> 
> I think would be a good idea to add a confirmation step for the user. I am worried about users losing their state/changes due to amending of the commit and have no way to go back.
> 
> We would also want the changes to go away from local state if user selected no for the made changes.

So, I agree that this would be ideal. But not sure if it is possible. Consider the case where the user has:

- Committed changes to `/some/path/foo` (1), and
- Uncommitted/unstaged changes to `/some/path/foo` (2)

Then a CPR resolve script runs, and it also changes `/some/path/foo`, creating a new state (3) for that file, which might include the unstaged changes plus the transformation made by the script (note that these are arbitrary scripts, and there is no way within `arc` to constrain what they might change, even if we can have some guidelines internally).

At this point, we can keep the version of `/some/path/foo` with all the changes in the local state (3) or revert to the fully committed version (1). But can't reasonably recover version (2).

One option is to `git stash` -> run resolve script -> prompt user, then:

- If user says "Yes" to applying the changes: add changed files -> amend commit -> `git stash pop`
- If user says "No": revert changed files to `HEAD` -> `git stash pop`

My worry there is what can happen if the process is interrupted, e.g. if the user `ctrl+C`s that last prompt. Then you end up with: a) strange changes to the local state, b) all of your local changes somehow stashed without your knowledge.

Another alternative is: If there are any uncommitted changes (other than `.gitignore`'d files), abort and tell the user that they might want to run the "resolve" script themselves.

What do you think?

**Comment by raviagarwal7** [comment]
from what I have seen arc asks user to stage and commit any un-staged changes when running arc diff. does the above workflow happen before or after CPR scripts run?

if arc diff's staging happens after CPR script run, then for every CPR script:
- Show confirmation to stage changes:
-- If yes, stage changes 
-- If no, fail (and ask user to get to a good state before running arc diff)
- Run CPR script
- If there are changes ask user for confirmation to stage changes:
-- If yes, stage the changes
-- If no, reset un-staged changes (this is fine to do since the changes came from CPR script)
  

we can do something similar as above if arc diff's staging happens before CPR script run.

**Comment by lazaroclapp** [comment]
> from what I have seen arc asks user to stage and commit any un-staged changes when running arc diff. does the above workflow happen before or after CPR scripts run?
> 
> if arc diff's staging happens after CPR script run, then for every CPR script:
> 
> * Show confirmation to stage changes:
>   -- If yes, stage changes
>   -- If no, fail (and ask user to get to a good state before running arc diff)
> * Run CPR script
> * If there are changes ask user for confirmation to stage changes:
>   -- If yes, stage the changes
>   -- If no, reset un-staged changes (this is fine to do since the changes came from CPR script)
> 
> we can do something similar as above if arc diff's staging happens before CPR script run.

Implemented this. See internal diff for how the flow looks in out current use case.

Sorry, accidentally made squashed this into a single commit ahead of time. Hope it doesn't make it much harder to review.

File: src/workflow/ArcanistDiffWorkflow.php [hunk 2]
```diff
@@ -1187,6 +1193,108 @@ private function shouldAmend() {
...
+   * @task checkPromptResolveScripts
+   */
+  private function runCheckPromptResolveScripts() {
+    // Skip this any time linting would be skipped, plus when explicitly
+    // passed --nocpr (meaning: run normal lint checks, but not CPR scripts).
+    if ($this->getArgument('nocpr') ||
+        $this->getArgument('nolint') ||
+        $this->getArgument('only') ||
+        $this->isRawDiffSource() ||
+        $this->getArgument('head')) {
```
[the remark this answers is not in the document]
Comment (lazaroclapp) [reply src/workflow/ArcanistDiffWorkflow.php]: We skip lint when running those two. My logic was that anything that disables lint should also disable CPR scripts. And then, additionally, you can use `--nocpr` if you want to disable only CPR scripts :)

[the remark this answers is not in the document]
Comment (lazaroclapp) [reply src/workflow/ArcanistDiffWorkflow.php]: Merging now based on the "approved" status, I can always open another PR if the above doesn't make sense. Thanks!

**Pull request merged**